### PR TITLE
plan: support pushing agg past union all.

### PR DESCRIPTION
**Pull request by hanfei1991**
support pushing agg past union all. This algo is very simple.
@coocood @zimulala @shenli @winoros @XuHuaiyu PTAL

### CODE REVIEW

File: plan/aggregation_push_down.go [hunk 1]
```diff
@@ -269,6 +246,61 @@ func (a *aggPushDownSolver) checkAnyCountAndSum(aggFuncs []expression.Aggregatio
...
+		firstRow := expression.NewAggFunction(ast.AggFuncFirstRow, []expression.Expression{gbyCol.Clone()}, false)
+		newAggFuncs = append(newAggFuncs, firstRow)
+		schema = append(schema, gbyCol.Clone().(*expression.Column))
+	}
+	agg.AggFuncs = newAggFuncs
+	agg.SetSchema(schema)
+	return agg
+}
+
+func (a *aggPushDownSolver) pushAggCrossUnion(agg *Aggregation, oldSchema expression.Schema, p LogicalPlan) LogicalPlan {
```
Comment (coocood) [inline plan/aggregation_push_down.go]: `oldSchema` -> `unionSchema`
`p` -> `unionChild`

File: plan/aggregation_push_down.go [hunk 2]
```diff
@@ -295,6 +327,17 @@ func (a *aggPushDownSolver) aggPushDown(p LogicalPlan) {
 				rChild.SetParents(join)
 				join.SetSchema(append(lChild.GetSchema().Clone(), rChild.GetSchema().Clone()...))
 			}
+		} else if union, ok1 := child.(*Union); ok1 {
+			pushedAgg := a.makeNewAgg(agg.AggFuncs, agg.groupByCols)
+			oldSchema := union.schema
+			union.SetSchema(pushedAgg.schema)
```
Comment (coocood) [inline plan/aggregation_push_down.go]: Maybe update union schema after push aggregate to children is better?
So we don't to define `oldSchema`.

**Comment by coocood** [comment]
LGTM

File: executor/executor_test.go [hunk 3]
```diff
@@ -485,6 +485,23 @@ func (s *testSuite) TestUnion(c *C) {
...
+	tk.MustExec("insert t1 values (1, 2)")
+	tk.MustExec("drop table if exists t2")
+	tk.MustExec("create table t2 (c int, d int)")
+	tk.MustExec("insert t2 values (1, 3)")
+	tk.MustExec("insert t2 values (1, 1)")
+	tk.MustExec("drop table if exists t3")
+	tk.MustExec("create table t3 (c int, d int)")
+	tk.MustExec("insert t3 values (3, 2)")
+	tk.MustExec("insert t3 values (4, 3)")
+
```
Comment (zimulala) [inline executor/executor_test.go]: Remove this empty line.

File: plan/aggregation_push_down.go [hunk 4]
```diff
@@ -269,6 +246,61 @@ func (a *aggPushDownSolver) checkAnyCountAndSum(aggFuncs []expression.Aggregatio
...
+	}
+	newAgg.collectGroupByColumns()
+	newAgg.SetChildren(unionChild)
+	unionChild.SetParents(newAgg)
+	return newAgg
+}
+
 // aggPushDown tries to push down aggregate functions to join paths.
 func (a *aggPushDownSolver) aggPushDown(p LogicalPlan) {
 	if agg, ok := p.(*Aggregation); ok {
```
Comment (hanfei1991) [inline plan/aggregation_push_down.go]: we shouldn't return , instead, we should go to line 342

**Comment by zimulala** [comment]
LGTM

**Pull request merged**